holopy/scattering/imageformation.py

```diff
@@ -0,0 +1,233 @@
+                illum_polarization=schema.illum_polarization)
+            )
+        phase = np.exp(-1j * wavevector * scatterer.center[2])
+        scattered_field *= phase
+        return scattered_field
+
+    def _pack_field_into_xarray(self, scattered_field, schema):
+        """Packs the numpy.ndarray, shape (N, 3) ``scattered_field`` into
+        an xr.DataArray, shape (N, 3). This function needs to pack the
+        fields [flat or point, vector], with the coordinates the
+        same as that of the schema."""
+        flattened_schema = flat(schema)  # now either point or flat
+        point_or_flat = self._is_detector_view_point_or_flat(flattened_schema)
+        coords = {
+            key: (point_or_flat, val.values)
+            for key, val in flattened_schema[point_or_flat].coords.items()}
+
+        coords.update(
+            {point_or_flat: flattened_schema[point_or_flat],
+             vector: ['x', 'y', 'z']})
+        scattered_field = xr.DataArray(
+            scattered_field, dims=[point_or_flat, vector], coords=coords,
+            attrs=schema.attrs)
+        return scattered_field
+
+    def _pack_scattering_matrix_into_xarray(
+            self, scat_matrs, r_theta_phi, schema):
+        flattened_schema = flat(schema)
+        point_or_flat = self._is_detector_view_point_or_flat(flattened_schema)
+        dims = [point_or_flat, 'Epar', 'Eperp']
+
+        coords = {point_or_flat: flattened_schema.coords[point_or_flat]}
+        coords.update({
+            'r': (point_or_flat, r_theta_phi[0]),
+            'theta': (point_or_flat, r_theta_phi[1]),
+            'phi': (point_or_flat, r_theta_phi[2]),
+            'Epar': ['S2', 'S3'],
+            'Eperp': ['S4', 'S1'],
+            })
+
+        packed = xr.DataArray(
+            scat_matrs, dims=dims, coords=coords, attrs=schema.attrs)
+        return packed
+
+    @classmethod
+    def _is_detector_view_point_or_flat(cls, detector_view):
+        detector_dims = detector_view.dims
+        if 'flat' in detector_dims:
+            point_or_flat = 'flat'
+        elif 'point' in detector_dims:
+            point_or_flat = 'point'
+        else:
+            msg = ("xarray `detector_view` is not in the form of a 1D list " +
+                   "of coordinates. Call ``flat`` first.")
+            raise ValueError(msg)
+        return point_or_flat
+
+    def _transform_to_desired_coordinates(self, detector, origin, wavevec=1):
+        if hasattr(detector, 'theta') and hasattr(detector, 'phi'):
+            original_coordinate_system = 'spherical'
+            original_coordinate_values = [
+                (detector.r.values * wavevec if hasattr(detector, 'r')
+                    else np.full(detector.theta.values.shape, np.inf)),
+                detector.theta.values,
+                detector.phi.values,
+                ]
+        else:
+            original_coordinate_system = 'cartesian'
+            f = flat(detector)  # 1.6 ms
```

Comment:
I don't think it's possible to pre-compute the coordinate transformation, since the cylindrical and spherical coordinates are usually defined with respect to the scatterer center. So changing the scatterer's position (as is done in frequently an inference calculation) would require re-computing the coordinate transformation. 

Pre-computing would help for multi-wavelength calculations, but would only give a ~2x speed-up. 

I think a better option is to step away from xarray earlier in the scattering theory calculation. Most of the time in this `_transform_to_desired_coordinates` function is actually spent just grabbing the coordinates from the xarray, not transforming them per se. One of my ulterior motives in doing this refactor was to make it easier to do this down the line.

Reply:
You're right, I obviously didn't think about this too much. The coordinate transforms should be really fast...